### PR TITLE
Fix #2072: erasure and refchecks

### LINKED ISSUE #2072: Concrete trait member should take precedence over deferred method
```scala
trait T { lazy val x: String = "foo" }
abstract class AC extends T { def x: Any }
```

compiles to:
```java
public abstract class AC implements T {
    public AC() { super.x();   }

    public abstract Object x(); // where did my lazy val go?
}

public interface T {
    default public void $init$() { }

    default public String x() {
        return "foo";
    }
}
```

See https://issues.scala-lang.org/browse/SI-10223

### PR DESCRIPTION
**Pull request by DarkDimius**

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/transform/TypeTestsCasts.scala [hunk 1]
```diff
@@ -96,23 +96,30 @@ trait TypeTestsCasts {
         /** Transform isInstanceOf OrType
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/transform/TypeTestsCasts.scala]: Not specific to OrType anymore

File: compiler/src/dotty/tools/dotc/transform/TypeTestsCasts.scala [hunk 2]
```diff
@@ -96,23 +96,30 @@ trait TypeTestsCasts {
...
          *
          *    expr.isInstanceOf[A | B]  ~~>  expr.isInstanceOf[A] | expr.isInstanceOf[B]
+         *    expr.isInstanceOf[A & B]  ~~>  expr.isInstanceOf[A] & expr.isInstanceOf[B]
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/transform/TypeTestsCasts.scala]: Would you mind moving this commit to a separate PR? It's not related to fixing #2072 and I'd like to get it in soon because https://github.com/lampepfl/dotty/pull/2082 depends on it.

File: compiler/src/dotty/tools/dotc/transform/Erasure.scala [hunk 3]
```diff
@@ -642,7 +647,22 @@ object Erasure extends TypeTestsCasts{
...
-                    sym => makeBridgeDef(member, sym)(ctx)
+                    sym =>
+                      if (member.symbol.is(Flags.Deferred)) {
+                        if (!sym.is(Flags.Deferred)) {
+                          // sometimes we need to forward an abstract method in a subclass to a non-abstract method in parent.
+                          // see https://github.com/lampepfl/dotty/issues/2072 for illustrations
+
+                          val d = member.symbol.copySymDenotation(initFlags = member.symbol.flags.&~(Flags.Deferred))
+                          d.installAfter(ctx.erasurePhase.asInstanceOf[Erasure])
+                          // Dark magic. Don't look here.
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/transform/Erasure.scala]: This is not a helpful comment. Someone might need to look at this eventually and will be very confused. Also do you need both `installAfter` and `transformAfter` ?

Comment (DarkDimius) [inline compiler/src/dotty/tools/dotc/transform/Erasure.scala]: > Also do you need both installAfter and transformAfter ?

Yes. The comment here would entirely repeat all the history of transformAfter. I don't think it's worth simplifying it here. To understand why it's needed here one need to read the entire description of 979ee0ff41c2a957b22e886ea43ad4f070a9777d which introduces transformAfter

Comment (smarter) [inline compiler/src/dotty/tools/dotc/transform/Erasure.scala]: Then the comment can say "See the commit message of 979ee0f for more information."

File: compiler/src/dotty/tools/dotc/transform/TypeTestsCasts.scala (hunk 2 above)
Comment (smarter) [inline compiler/src/dotty/tools/dotc/transform/TypeTestsCasts.scala]: An identical commit has now been merged as part of https://github.com/lampepfl/dotty/pull/2082
